From: Folia Compatibility Patch
Subject: [PATCH 0018] ExecutionTask: Migrate ExecutionTask to SchedulerUtils and update AdvancedWeapon attributes

AE PATCH REASON: Core effect engine uses synchronous waits and incompatible attributes
AE PATCH FIX: Migrate ExecutionTask to SchedulerUtils and update AdvancedWeapon attributes
AE PATCH RISK: HIGH
AE PATCH EXCLUDES: net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.class, net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask$*.class

---
diff --git a/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java b/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java
index e6c3353..2ce6e68 100644
--- a/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java
+++ b/net/advancedplugins/ae/impl/effects/effects/actions/execution/ExecutionTask.java
@@ -4,7 +4,6 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.LinkedList;
 import java.util.List;
-import lombok.Generated;
 import net.advancedplugins.ae.impl.effects.api.EffectsActivateEvent;
 import net.advancedplugins.ae.impl.effects.api.EffectsActivatedEvent;
 import net.advancedplugins.ae.impl.effects.armorutils.ArmorType;
@@ -12,7 +11,6 @@ import net.advancedplugins.ae.impl.effects.effects.EffectsHandler;
 import net.advancedplugins.ae.impl.effects.effects.abilities.AdvancedAbility;
 import net.advancedplugins.ae.impl.effects.effects.actions.ActionExecution;
 import net.advancedplugins.ae.impl.effects.effects.actions.ActionExecutionBuilder;
-import net.advancedplugins.ae.impl.effects.effects.actions.execution.ExecutionTask.1;
 import net.advancedplugins.ae.impl.effects.effects.actions.handlers.DamageHandler;
 import net.advancedplugins.ae.impl.effects.effects.effects.AdvancedEffect;
 import net.advancedplugins.ae.impl.effects.effects.mechanics.targets.TargetResults;
@@ -21,7 +19,6 @@ import net.advancedplugins.ae.impl.effects.effects.variables.DynamicVariable;
 import net.advancedplugins.ae.impl.effects.effects.variables.Variables;
 import net.advancedplugins.ae.impl.utils.ASManager;
 import net.advancedplugins.ae.impl.utils.LocalLocation;
-import net.advancedplugins.ae.impl.utils.SchedulerUtils;
 import net.advancedplugins.ae.impl.utils.hooks.HookPlugin;
 import net.advancedplugins.ae.impl.utils.hooks.HooksHandler;
 import net.advancedplugins.ae.impl.utils.hooks.plugins.MythicMobsHook;
@@ -46,216 +43,227 @@ public class ExecutionTask {
    private TargetResults targetResults;
    private final DamageHandler damageHandler = new DamageHandler();
 
-   public ExecutionTask(AdvancedAbility var1, ActionExecution var2, ActionExecutionBuilder var3) {
-      this.ability = var1;
-      this.actionExecution = var2;
-      this.builder = var3;
-      this.location = var3.getMain().getLocation();
+   public ExecutionTask(AdvancedAbility ability, ActionExecution actionExecution, ActionExecutionBuilder builder) {
+      this.ability = ability;
+      this.actionExecution = actionExecution;
+      this.builder = builder;
+      this.location = builder.getMain().getLocation();
    }
 
    public void init() {
-      if (!this.soulboundOnly || !this.actionExecution.getAllEffects().stream().noneMatch(var0 -> var0.equalsIgnoreCase("KEEP_ON_DEATH"))) {
-         ArrayList var1 = new ArrayList();
+      if (!this.soulboundOnly || !this.actionExecution.getAllEffects().stream().noneMatch(s -> s.equalsIgnoreCase("KEEP_ON_DEATH"))) {
+         List<LivingEntity> entitiesToAddFromEvent = new ArrayList<>();
          if (!this.sets && Bukkit.isPrimaryThread()) {
-            EffectsActivateEvent var2 = new EffectsActivateEvent(this.ability, this.builder.getMain(), this.builder.getOther(), this);
-            Bukkit.getPluginManager().callEvent(var2);
-            if (this.ability.getEffects().isEmpty() || var2.isCancelled()) {
+            EffectsActivateEvent effectsActivateEvent = new EffectsActivateEvent(this.ability, this.builder.getMain(), this.builder.getOther(), this);
+            Bukkit.getPluginManager().callEvent(effectsActivateEvent);
+            if (this.ability.getEffects().isEmpty() || effectsActivateEvent.isCancelled()) {
                return;
             }
 
-            if (var2.getOtherTargets() != null) {
-               var1.addAll(var2.getOtherTargets());
+            if (effectsActivateEvent.getOtherTargets() != null) {
+               entitiesToAddFromEvent.addAll(effectsActivateEvent.getOtherTargets());
             }
          }
 
-         for (String var3 : this.ability.getEffects()) {
-            ArrayList var4 = new ArrayList();
-            LinkedList var5 = new LinkedList();
-            var3 = this.actionExecution.parseVariables(var3);
-            this.updateMain(var5, var4, var3);
-            var5.addAll(var1);
-            if (var3.startsWith("WAIT")) {
+         for (String effect : this.ability.getEffects()) {
+            final List<Location> targetLocations = new ArrayList<>();
+            final LinkedList<LivingEntity> targetEntities = new LinkedList<>();
+            effect = this.actionExecution.parseVariables(effect);
+            this.updateMain(targetEntities, targetLocations, effect);
+            targetEntities.addAll(entitiesToAddFromEvent);
+            if (effect.startsWith("WAIT")) {
                this.wait = true;
             }
 
             if (this.targetResults.getEffect() != null) {
-               var3 = this.actionExecution.parseVariables(this.targetResults.getEffect());
+               effect = this.actionExecution.parseVariables(this.targetResults.getEffect());
             }
 
-            String[] var6 = var3.split(":");
-            if (var3.startsWith("WAIT")) {
-               this.currentWait = this.currentWait + ASManager.parseInt(var6[1], 0);
+            String[] baseEffectSplit = effect.split(":");
+            if (effect.startsWith("WAIT")) {
+               this.currentWait = this.currentWait + ASManager.parseInt(baseEffectSplit[1], 0);
             } else if (!this.wait) {
-               if (!var4.isEmpty()) {
-                  for (Location var12 : var4) {
-                     this.activate(var3, this.builder.getMain(), var12);
+               if (!targetLocations.isEmpty()) {
+                  for (Location loc : targetLocations) {
+                     this.activate(effect, this.builder.getMain(), loc);
                   }
                } else {
-                  for (LivingEntity var8 : var5) {
-                     this.activate(var3, var8, null);
+                  for (LivingEntity target : targetEntities) {
+                     this.activate(effect, target, null);
                   }
                }
             } else {
-               new 1(this, var4, var3, var5).runTaskLater(EffectsHandler.getInstance(), this.currentWait);
+               final String finalEffect = effect;
+               long delay = (long)this.currentWait < 1L ? 1L : (long)this.currentWait;
+               if (!targetLocations.isEmpty()) {
+                  for (Location loc : targetLocations) {
+                     org.bukkit.Bukkit.getRegionScheduler().runDelayed(ASManager.getInstance(), loc, task2 -> {
+                        ExecutionTask.this.builder.setDamageEventNotGoingToRun(true);
+                        ExecutionTask.this.activate(finalEffect, ExecutionTask.this.builder.getMain(), loc);
+                     }, delay);
+                  }
+               } else {
+                  for (LivingEntity target : targetEntities) {
+                     target.getScheduler().runDelayed(ASManager.getInstance(), task2 -> {
+                        ExecutionTask.this.builder.setDamageEventNotGoingToRun(true);
+                        ExecutionTask.this.activate(finalEffect, target, null);
+                     }, () -> {}, delay);
+                  }
+               }
             }
          }
       }
    }
 
-   private void activate(String var1, LivingEntity var2, Location var3) {
-      ASManager.debug("§d[ExecutionTask] start of " + var1 + " parsing+");
-      String var4 = DynamicVariable.parseThroughCustomVariables(
-         Variables.replaceVariables(var1, this.builder.getAttacker(), this.builder.getVictim(), this.actionExecution)
+   private void activate(String initialEffect, LivingEntity ent, Location loc) {
+      ASManager.debug("§d[ExecutionTask] start of " + initialEffect + " parsing+");
+      String effect = DynamicVariable.parseThroughCustomVariables(
+         Variables.replaceVariables(initialEffect, this.builder.getAttacker(), this.builder.getVictim(), this.actionExecution)
       );
-      var4 = EffectsHandler.getVariablesHandler().parseEffectLine(var4, this.builder.getMain(), this);
-      var4 = EffectsHandler.getFunctionsHandler().parseEffectLine(var4, this.builder.getMain(), this);
-      var4 = EffectsHandler.getPointersHandler().parseEffectLine(var4, this.builder.getType(), this);
-      if (!var4.isEmpty() && !var4.contains("$skip")) {
-         String[] var5 = var4.replaceAll(" ", "").split(":");
-         String[] var6 = Arrays.copyOfRange(var5, 1, var5.length);
-         AdvancedEffect var7 = EffectsHandler.getEffectStorage().getEffect(var5[0].replaceAll(" ", ""));
-         if (var7 == null) {
+      effect = EffectsHandler.getVariablesHandler().parseEffectLine(effect, this.builder.getMain(), this);
+      effect = EffectsHandler.getFunctionsHandler().parseEffectLine(effect, this.builder.getMain(), this);
+      effect = EffectsHandler.getPointersHandler().parseEffectLine(effect, this.builder.getType(), this);
+      if (!effect.isEmpty() && !effect.contains("$skip")) {
+         String[] effectSplit = effect.replaceAll(" ", "").split(":");
+         String[] arguments = Arrays.copyOfRange(effectSplit, 1, effectSplit.length);
+         AdvancedEffect advancedEffect = EffectsHandler.getEffectStorage().getEffect(effectSplit[0].replaceAll(" ", ""));
+         if (advancedEffect == null) {
             this.reportIssue(
-               var4,
-               "Failed to activate effects as advancedEffect is null or invalid: '" + var5[0].replaceAll(" ", "") + "'",
-               var1,
+               effect,
+               "Failed to activate effects as advancedEffect is null or invalid: '" + effectSplit[0].replaceAll(" ", "") + "'",
+               initialEffect,
                "trigger:" + this.builder.getType(),
                "entity:" + this.builder.getMain().getName()
             );
          } else {
-            if (var7.hasStringArgument()) {
-               var5 = var4.split(":");
-               var6 = Arrays.copyOfRange(var5, 1, var5.length);
+            if (advancedEffect.hasStringArgument()) {
+               effectSplit = effect.split(":");
+               arguments = Arrays.copyOfRange(effectSplit, 1, effectSplit.length);
             }
 
-            if (var7.isBlockEffect() && var3 == null && this.builder.getEvent() instanceof BlockBreakEvent) {
-               var3 = this.builder.getBlock().getLocation();
+            if (advancedEffect.isBlockEffect() && loc == null && this.builder.getEvent() instanceof BlockBreakEvent) {
+               loc = this.builder.getBlock().getLocation();
             }
 
-            if (this.builder.getMain() instanceof Player && !this.builder.getMain().isOp() && var7.isExemptFromAC()) {
-               EffectsHandler.getAntiCheatHooks().forEach((var1x, var2x) -> var2x.exempt((Player)this.builder.getMain()));
+            if (this.builder.getMain() instanceof Player && !this.builder.getMain().isOp() && advancedEffect.isExemptFromAC()) {
+               EffectsHandler.getAntiCheatHooks().forEach((key, value) -> value.exempt((Player)this.builder.getMain()));
             }
 
             if (this.builder.getEvent() instanceof EntityDamageByEntityEvent) {
-               EntityDamageByEntityEvent var8 = (EntityDamageByEntityEvent)this.builder.getEvent();
-               if (HooksHandler.isEnabled(HookPlugin.MYTHICMOBS) && MythicMobsHook.getIgnoreEnchantsMobs().contains(var8.getDamager())) {
+               EntityDamageByEntityEvent event = (EntityDamageByEntityEvent)this.builder.getEvent();
+               if (HooksHandler.isEnabled(HookPlugin.MYTHICMOBS) && MythicMobsHook.getIgnoreEnchantsMobs().contains(event.getDamager())) {
                   return;
                }
             }
 
             ASManager.debug(
                "§d§l(!) "
-                  + var4
+                  + effect
                   + " activating "
-                  + (var3 != null ? "location based " + new LocalLocation(var3).getEncode() : "entity based " + (var2 != null ? var2.getType().name() : "null"))
+                  + (loc != null ? "location based " + new LocalLocation(loc).getEncode() : "entity based " + (ent != null ? ent.getType().name() : "null"))
             );
-
             try {
-               if (var3 != null) {
-                  if (var7.executeEffect(this, var3, var6)) {
+               if (loc != null) {
+                  if (advancedEffect.executeEffect(this, loc, arguments)) {
                      if (!Bukkit.isPrimaryThread()) {
-                        SchedulerUtils.runTask(() -> {
-                           EffectsActivatedEvent var2x = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                           Bukkit.getPluginManager().callEvent(var2x);
+                        Bukkit.getRegionScheduler().execute(ASManager.getInstance(), loc, () -> {
+                           EffectsActivatedEvent effectsActivatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), ent, this);
+                           Bukkit.getPluginManager().callEvent(effectsActivatedEvent);
                         });
                      } else {
-                        EffectsActivatedEvent var14 = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                        Bukkit.getPluginManager().callEvent(var14);
+                        EffectsActivatedEvent effectsActivatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), ent, this);
+                        Bukkit.getPluginManager().callEvent(effectsActivatedEvent);
                      }
                   } else {
-                     this.reportIssue(var4, "activating entity-based effect without a valid target (e.g. @Attacker)", this.ability.getName());
+                     this.reportIssue(effect, "activating entity-based effect without a valid target (e.g. @Attacker)", this.ability.getName());
                   }
-               } else if (!var7.executeEffect(this, var2, var6)) {
-                  this.reportIssue(var4, "activating location-based effect without a valid target (e.g. @Block)", this.ability.getName());
+               } else if (!advancedEffect.executeEffect(this, ent, arguments)) {
+                  this.reportIssue(effect, "activating location-based effect without a valid target (e.g. @Block)", this.ability.getName());
                } else {
                   if (!Bukkit.isPrimaryThread()) {
-                     SchedulerUtils.runTask(() -> {
-                        EffectsActivatedEvent var2x = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                        Bukkit.getPluginManager().callEvent(var2x);
-                     });
+                     LivingEntity targetEnt = ent != null ? ent : this.builder.getMain();
+                     targetEnt.getScheduler().run(ASManager.getInstance(), task2 -> {
+                        EffectsActivatedEvent effectsActivatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), ent, this);
+                        Bukkit.getPluginManager().callEvent(effectsActivatedEvent);
+                     }, null);
                   } else {
-                     EffectsActivatedEvent var15 = new EffectsActivatedEvent(this.ability, this.builder.getMain(), var2, this);
-                     Bukkit.getPluginManager().callEvent(var15);
+                     EffectsActivatedEvent effectsActivatedEvent = new EffectsActivatedEvent(this.ability, this.builder.getMain(), ent, this);
+                     Bukkit.getPluginManager().callEvent(effectsActivatedEvent);
                   }
 
                   if (this.getBuilder().isPermanent() && this.getBuilder().isRemoved()) {
-                     SchedulerUtils.runTaskLater(
-                        () -> {
-                           if (var2 != null && !var2.isDead() && ASManager.isOnline(var2)) {
-                              ArmorType.getArmorContents(var2)
-                                 .forEach((var1x, var2x) -> ArmorWearTrigger.getArmorWearTrigger().runCheck(var2, var2x, var1x, false, false));
+                     if (ent != null) {
+                        ent.getScheduler().runDelayed(ASManager.getInstance(), task2 -> {
+                           if (!ent.isDead() && ASManager.isOnline(ent)) {
+                              ArmorType.getArmorContents(ent)
+                                 .forEach((armorType, itemStack) -> ArmorWearTrigger.getArmorWearTrigger().runCheck(ent, itemStack, armorType, false, false));
                            }
-                        },
-                        2L
-                     );
+                        }, () -> {}, 2L);
+                     }
                   }
                }
             } catch (Exception var9) {
                var9.printStackTrace();
-               this.reportIssue(var4, "Error while activating effect");
+               this.reportIssue(effect, "Error while activating effect");
             }
          }
       }
    }
 
-   public void reportIssue(String var1, String... var2) {
-      for (Player var4 : Bukkit.getOnlinePlayers()) {
-         if (var4.hasPermission("ae.admin")) {
-            var4.sendMessage(Text.modify("&4Failed to activate effect '&f" + var1 + "'"));
-            var4.sendMessage(Text.modify("&cAdditional information: &7" + String.join(", ", var2)));
+   public void reportIssue(String effect, String... args) {
+      for (Player p : Bukkit.getOnlinePlayers()) {
+         if (p.hasPermission("ae.admin")) {
+            p.sendMessage(Text.modify("&4Failed to activate effect '&f" + effect + "'"));
+            p.sendMessage(Text.modify("&cAdditional information: &7" + String.join(", ", args)));
          }
       }
 
-      ASManager.getInstance().getLogger().warning(Text.modify("&4Failed to activate effect '&f" + var1 + "'"));
-      ASManager.getInstance().getLogger().warning(Text.modify("&cAdditional information: &7" + String.join(", ", var2)));
+      ASManager.getInstance().getLogger().warning(Text.modify("&4Failed to activate effect '&f" + effect + "'"));
+      ASManager.getInstance().getLogger().warning(Text.modify("&cAdditional information: &7" + String.join(", ", args)));
    }
 
-   private void updateMain(LinkedList<LivingEntity> var1, List<Location> var2, String var3) {
-      var1.clear();
-      var2.clear();
-      this.targetResults = EffectsHandler.getTargetHandler().handleTargets(var3, this);
+   private void updateMain(LinkedList<LivingEntity> targetEntities, List<Location> targetLocations, String effect) {
+      targetEntities.clear();
+      targetLocations.clear();
+      this.targetResults = EffectsHandler.getTargetHandler().handleTargets(effect, this);
       if (this.targetResults.getTargetList() != null) {
-         var1.addAll(this.targetResults.getTargetList());
+         targetEntities.addAll(this.targetResults.getTargetList());
       }
 
       if (this.targetResults.getTargetLocations() != null) {
-         var2.addAll(this.targetResults.getTargetLocations());
+         targetLocations.addAll(this.targetResults.getTargetLocations());
       }
 
-      var2 = ASManager.removeDuplicateLocations(var2);
+      targetLocations = ASManager.removeDuplicateLocations(targetLocations);
    }
 
-   public ExecutionTask asSets(boolean var1) {
-      this.sets = var1;
+   public ExecutionTask asSets(boolean sets) {
+      this.sets = sets;
       return this;
    }
 
-   public ExecutionTask setLocation(Location var1) {
-      this.location = var1;
+   public ExecutionTask setLocation(Location location) {
+      this.location = location;
       return this;
    }
 
-   public ExecutionTask soulboundOnly(boolean var1) {
-      this.soulboundOnly = var1;
+   public ExecutionTask soulboundOnly(boolean b) {
+      this.soulboundOnly = b;
       return this;
    }
 
-   @Generated
    public ActionExecution getActionExecution() {
       return this.actionExecution;
    }
 
-   @Generated
    public ActionExecutionBuilder getBuilder() {
       return this.builder;
    }
 
-   @Generated
    public AdvancedAbility getAbility() {
       return this.ability;
    }
 
-   @Generated
    public DamageHandler getDamageHandler() {
       return this.damageHandler;
    }

